Documentation/designs/status-subresource.md

```diff
@@ -0,0 +1,436 @@
+# Status subresource for Prometheus operator CRDs
+
+## Summary
+
+Though some of the custom resource definitions expose already a `Status`
+subresource, the Prometheus operator never updates the status subresource and
+it is only possible to retrieve the information via the custom API exposed by
+the operator’s web service. This makes it harder than necessary for users to
+know if the declared resources are ready and to understand why if they aren’t.
+
+## Goals
+
+* Define the structure of the status subresource for the custom resource definitions managed by the Prometheus operator.
+* Define how the operator would reconcile the status subresource.
+
+## Non-goals
+
+* Extend the status subresource beyond what the operator can infer from the core Kubernetes resources (e.g. we don't want to correlate the service/pod monitors with the status of the targets as reported by Prometheus, at least for now). This would require a back-channel communication from the operands to the operator which doesn't exist today.
+* Emit events on resource updates. Once the operator implements status subresources, it seems a natural evolution to generate events on status changes but this isn't in the scope of this proposal.
+
+## Background
+
+The status subresource is a well-defined concept in Kubernetes:
+* [Kubernetes documentation](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#status-subresource) about custom resource definitions.
+* [Kubernetes API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties).
+* [kubebuilder](https://book-v1.book.kubebuilder.io/basics/status_subresource.html) documentation.
+* [OperatorSDK](https://sdk.operatorframework.io/docs/building-operators/golang/advanced-topics/#manage-cr-status-conditions) documentation.
+
+As much as possible, the proposal follows the recommendations established by the Kubernetes community.
+
+The feature is tracked in [#3335](https://github.com/prometheus-operator/prometheus-operator/issues/3335).
+
+## API
+
+### Prometheus
+
+The Prometheus CRD has a `Status` subresource that exposes the following fields:
+* `Paused`
+* `Replicas`
+* `UpdatedReplicas`
+* `AvailableReplicas`
+* `UnavailableReplicas`
```

Comment:
Is `UnavailableReplicas` the same as `Replicas - AvailableReplicas`? If so I don't think we need it.